code/espurna/rpnrules.cpp

```diff
@@ -124,52 +140,219 @@ void _rpnBrokerStatus(const String& topic, unsigned char id, unsigned int value)
+        for (int index = 0; index < 2; ++index) {
+            refs[index] = (raw[index] == recent.raw) ? &recent : nullptr;
+        }
+    }
+
+    if ((refs[0] == nullptr) || (refs[1] == nullptr)) {
+        return rpn_operator_error::CannotContinue;
     }
-    return true;
+
+    // purge codes to avoid matching again on the next rules run
+    // 'triggered' would be kind of quirky here, since we don't limit ourselves with any window
+    if ((millis() - refs[0]->last) > (millis() - refs[1]->last)) {
+        _rfb_codes.remove_if([&refs](rpn_rfbridge_code& code) {
+            return (refs[0] == &code) || (refs[1] == &code);
+        });
+        return rpn_operator_error::Ok;
+    }
+
+    return rpn_operator_error::CannotContinue;
+}
+
+rpn_error _rpnRfbMatcher(rpn_context& ctxt) {
+    rpn_value code;
+    rpn_stack_pop(ctxt, code);
+
+    rpn_value hits;
+    rpn_stack_pop(ctxt, hits);
+
+    String raw_code = code.toString();
+    auto result = std::find_if(_rfb_codes.begin(), _rfb_codes.end(), [&raw_code](const rpn_rfbridge_code& code) {
+        return code.raw == raw_code;
+    });
+
+    if (result == _rfb_codes.end()) {
+        return rpn_operator_error::CannotContinue;
+    }
+
+    // only process recent codes, ignore when rule is processing outside of this small window
+    if (millis() - (*result).last >= 2000) {
+        return rpn_operator_error::CannotContinue;
+    }
+
+    // hits == 1 is a single click, hits == 5 is long click
+    // we sort-of can distinguish single and double via timestamp, but it has a **very** unreliable timing
+    if ((*result).hits > hits.toUint()) {
+        return rpn_operator_error::CannotContinue;
+    }
+
+    // avoid re-triggering immediatly based on something running rules engine again
+    if ((*result).triggered && (millis() - (*result).triggered <= 1000)) {
+        return rpn_operator_error::CannotContinue;
+    }
+    (*result).triggered = millis();
+
+    return rpn_operator_error::Ok;
 }
 
+void _rpnBrokerRfbridgeCallback(const char* raw_code) {
+
+#if RFB_DIRECT
+    raw_code = (raw_code + strlen(raw_code) - 8);
+#else
+    raw_code = (raw_code + strlen(raw_code) - 6);
 #endif
 
+    // expire really old codes to avoid memory exhaustion
+    auto ts = millis();
+    auto old = std::remove_if(_rfb_codes.begin(), _rfb_codes.end(), [ts](rpn_rfbridge_code& code) {
+        return (ts - code.last) >= 10000u;
+    });
+
+    if (old != _rfb_codes.end()) {
+        _rfb_codes.erase(old, _rfb_codes.end());
+    }
+
+    auto result = std::find_if(_rfb_codes.begin(), _rfb_codes.end(), [raw_code](rpn_rfbridge_code& code) {
+        return code.raw == raw_code;
+    });
+
+    if (result != _rfb_codes.end()) {
+        // we also need to reset hits at a certain point to allow repeats to go through
+        // number is arbitrary time, just about 3-4 times more than it takes for a code to repeat when holding
+        if (millis() - (*result).last >= 2000u) {
```

Comment:
Some hard-coded values. We reset 'hits' counter when code was last received more than 2 seconds ago